File: projects/order-creation-queries/app/src/main/java/com/robinfood/app/controllers/configuration/stores/StoreController.java
package com.robinfood.app.controllers.configuration.stores;

import com.robinfood.app.usecases.getstores.IGetStoresUseCase;
import com.robinfood.app.usecases.getstoresbycountryid.IGetStoresByCountryIdUseCase;
import com.robinfood.core.dtos.ApiResponseDTO;
import com.robinfood.core.dtos.configuration.StoreWithIdAndNameDTO;
import com.robinfood.core.enums.Result;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.Objects;

import static com.robinfood.core.constants.APIConstants.STORES_V1;
import static com.robinfood.core.constants.APIConstants.STORES;

@RequestMapping(STORES_V1)
@RestController
@Slf4j
public class StoreController implements IStoresController {

    private final IGetStoresUseCase getStoresUseCase;
    private final IGetStoresByCountryIdUseCase getStoresByCountryIdUseCase;

    public StoreController(
            IGetStoresUseCase getConfigStoreUseCase,
            IGetStoresByCountryIdUseCase getStoresByCountryIdUseCase
    ) {
        this.getStoresUseCase = getConfigStoreUseCase;
        this.getStoresByCountryIdUseCase = getStoresByCountryIdUseCase;
    }

    @Override
    @GetMapping(STORES)
    public ResponseEntity<ApiResponseDTO<List<StoreWithIdAndNameDTO>>> invoke(Long countryId) {

        log.info("Receiving request get Stores with country id: {}", countryId);

        Result<List<StoreWithIdAndNameDTO>> getStores;

        if (Objects.nonNull(countryId)) {
            getStores = getStoresByCountryIdUseCase.invoke(countryId);
        } else {
            getStores = getStoresUseCase.invoke();
        }

        ApiResponseDTO<List<StoreWithIdAndNameDTO>> apiResponseDTO;

        HttpStatus httpStatus;

        if (getStores instanceof Result.Error) {
            httpStatus = ((Result.Error) getStores).getHttpStatus();
            apiResponseDTO = new ApiResponseDTO<>(
                    ((Result.Error) getStores).getException().getLocalizedMessage(),
                    httpStatus
            );
        } else {
            httpStatus = HttpStatus.OK;
            apiResponseDTO = new ApiResponseDTO<>(
                    ((Result.Success<List<StoreWithIdAndNameDTO>>) getStores).getData(),
                    httpStatus
            );
        }

        return new ResponseEntity<>(apiResponseDTO, httpStatus);
    }
}

File: projects/posv2/app/src/main/kotlin/com/robinfood/app/controllers/store/StoreController.kt
package com.robinfood.app.controllers.store

import com.robinfood.app.usecases.getstoredeliveryplatforms.IGetStoreDeliveryPlatformsUseCase
import com.robinfood.core.constants.APIConstants.API_V1
import com.robinfood.core.constants.APIConstants.STORE
import com.robinfood.core.dtos.ApiResponseDTO
import com.robinfood.core.dtos.store.StoreDeliveryPlatformDTO
import com.robinfood.core.enums.Result

import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.server.ResponseStatusException
import javax.servlet.http.HttpServletRequest

/**
 * Implementation of 'IStoreController'
 */
@RestController
@RequestMapping(API_V1 + STORE)
class StoreController(
        private val getStoreDeliveryPlatformsUseCase: IGetStoreDeliveryPlatformsUseCase
) : IStoreController {

    @GetMapping("/{storeId}/delivery-platforms")
    override suspend fun getDeliveryPlatforms(
            httpServletRequest: HttpServletRequest,
            @RequestParam(value = "page") page: Int?,
            @RequestParam(value = "size") size: Int?,
            @PathVariable storeId: Long
    ): ResponseEntity<ApiResponseDTO<List<StoreDeliveryPlatformDTO>>> {
        val token = httpServletRequest.getHeader(HttpHeaders.AUTHORIZATION)
        return when (
            val resultStoreDeliveryPlatforms = getStoreDeliveryPlatformsUseCase.invoke(
                    page,
                    size,
                    storeId,
                    token
            )
        ) {
            is Result.Success -> {
                var responseStatus = HttpStatus.OK
                var responseMessage = "Store delivery platforms retrieved successfully"
                if (resultStoreDeliveryPlatforms.data.isEmpty()) {
                    responseMessage = "There are no platforms assigned to the store"
                    responseStatus = HttpStatus.NO_CONTENT
                }
                ResponseEntity(
                        ApiResponseDTO(
                                resultStoreDeliveryPlatforms.data,
                                responseMessage
                        ),
                        responseStatus
                )
            }
            is Result.Error -> throw ResponseStatusException(
                    HttpStatus.BAD_REQUEST,
                    resultStoreDeliveryPlatforms.exception.localizedMessage
            )
        }
    }
}
